src/cmap/conn/mod.rs

```diff
@@ -228,26 +212,130 @@ impl Connection {
             .into()
         })
     }
+
+    /// Close this connection, emitting a `ConnectionClosedEvent` with the supplied reason.
+    pub(super) fn close_and_drop(mut self, reason: ConnectionClosedReason) {
+        self.close(reason);
+    }
+
+    /// Close this connection, emitting a `ConnectionClosedEvent` with the supplied reason.
+    fn close(&mut self, reason: ConnectionClosedReason) {
+        self.detatch_from_pool();
+        if let Some(ref handler) = self.handler {
+            handler.handle_connection_closed_event(self.closed_event(reason));
+        }
+    }
+
+    /// Removes the weak reference to the parent pool from this connection.
+    /// If this connection is dropped, it will not emit any events.
+    fn detatch_from_pool(&mut self) {
```

Comment:
I'm not super fond of this method name, as it seems weird to be calling something called `detach_from_pool` from `mark_checked_in`. I'm thinking a name that describes it at a lower-level might be better, e.g. `remove_pool_reference`. That being said, I'm not convinced this really needs to be a separate method; calling `self.pool.take()` in the two places internally doesn't seem like a big deal to me.